### PR TITLE
feat: Add HTML support in explanation

### PR DESCRIPTION
**Pull request by abodacs**
## Description
This PR adds HTML support in explanations within DnDv2

After:
screenshot-rendered-html

Supporting information
OpenCraft internal ticket : [BB-6630](https://tasks.opencraft.com/browse/BB-6630)

## Testing instructions

1. Checkout this branch in <devstack_parent>/src directory in your local setup
2. log in to Studio shell, switch to edxapp virtual env
3. Go to /edx/src/xblock-drag-and-drop-v2 and run ```pip install -e .```
5. [enable the Drag and Drop XBlock in Studio](https://github.com/open-craft/xblock-drag-and-drop-v2/#enabling-in-studio)
5. Go to Studio UI and create a new unit
6. Select Problem > Advanced > Drag and Drop, to create a v2 drag and drop XBlock
7. Click Edit and change mode to Assessment
8. Scroll down to Explanation Text like `Test HTML 12m<sup>2</sup>`
9. Add some text in the text area
10. Click Continue twice and Save
11. Click on View Live Version
12. Ensure there is no extra line as shown above
13. Do as many attempts as required till "Show Answer" button is active
14. Click on Show Answer
15. Ensure Explanation is rendered like 'Test HTML  12m<sup>2</sup>'

## Deadline
"None"

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @abodacs!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.


:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://openedx.org/cla)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.     If you've signed an agreement in the past, you may need to re-sign.  See     [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/)     for details.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @abodacs!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.


:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://openedx.org/cla)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.     If you've signed an agreement in the past, you may need to re-sign.  See     [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/)     for details.

**Comment by natabene** [comment]
@abodacs Thanks for the PR, we will be waiting for the CLA check to go green first.

File: drag_and_drop_v2/utils.py [hunk 1]
```diff
@@ -22,7 +21,7 @@ def ngettext_fallback(text_singular, text_plural, number):
         return text_plural
 
 
-def _clean_data(data):
+def clean_data(data):
```
Comment (tecoholic) [inline drag_and_drop_v2/utils.py:24]: @abodacs Can your kindly revert this and other related renames?

File: drag_and_drop_v2/drag_and_drop_v2.py [hunk 2]
```diff
@@ -620,8 +620,14 @@ def show_answer(self, data, suffix=''):
 
         answer = self._get_correct_state()
 
-        answer['explanation'] = self.data.get('explanation', '').strip()
-
+        explanation = self.data.get('explanation', '').strip()
```
Comment (tecoholic) [inline drag_and_drop_v2/drag_and_drop_v2.py:623]: @abodacs 
```suggestion
        explanation = _clean_data(self.data.get('explanation', '').strip())
```

**Review by tecoholic — CHANGES_REQUESTED** [review]
@abodacs Looks close to being done. Thank you for your patience. I have added a couple of suggestions. I will add my approval once those are addressed.

File: drag_and_drop_v2/utils.py (hunk 1 above)
Comment (abodacs) [reply drag_and_drop_v2/utils.py]: @tecoholic 
Done Thanks

**Review by tecoholic — APPROVED** [review]
@abodacs :+1:

- [x] I tested this: Entered HTML in the explanation box and verified that it renders properly in the window
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] Includes documentation - NA

File: drag_and_drop_v2/drag_and_drop_v2.py (hunk 2 above)
Comment (abodacs) [reply drag_and_drop_v2/drag_and_drop_v2.py:623]: I revered adding  ```_clean_data``` suggestion as it removed HTML tags

Comment (tecoholic) [reply drag_and_drop_v2/drag_and_drop_v2.py:623]: But it was working fine when I was testing it the other day. I will check its impact.

File: drag_and_drop_v2/drag_and_drop_v2.py [hunk 3]
```diff
@@ -620,8 +620,14 @@ def show_answer(self, data, suffix=''):
...
+        if hasattr(self.runtime, 'replace_urls'):
+            explanation = self.runtime.replace_urls(explanation)
+        if hasattr(self.runtime, 'replace_course_urls'):
+            explanation = self.runtime.replace_course_urls(explanation)
+        if hasattr(self.runtime, 'replace_jump_to_id_urls'):
```
Comment (Agrendalath) [inline drag_and_drop_v2/drag_and_drop_v2.py:628]: Why are we checking the attr existence for all these runtime methods?

File: drag_and_drop_v2/utils.py [hunk 4]
```diff
@@ -143,7 +142,6 @@ def apply_zone_migrations(self, zone):
         """
         migrations = (self._zone_v1_to_v2, self._zone_v2_to_v2p1)
         zone_id = zone.get('uid', zone.get('id'))
-
```
Comment (Agrendalath) [inline drag_and_drop_v2/utils.py:146]: We're not modifying this file in this PR. Please revert newline changes.

File: tests/integration/test_interaction_assessment.py [hunk 5]
```diff
@@ -474,13 +474,29 @@ def load_scenario(self, explanation_text: str, scenario_id: int):
         self._add_scenario(scenario_page_id, scenario_page_title, scenario_xml)
         self._page = self.go_to_page(scenario_page_title)
 
+    @staticmethod
+    def _get_explanation_html(explanation: str):
+        """
+        get generated html explanation
```
Comment (Agrendalath) [inline tests/integration/test_interaction_assessment.py:480]: Nit: the sentence in the docstring should start with an uppercase letter and end with a full stop.

File: tests/integration/test_interaction_assessment.py [hunk 6]
```diff
@@ -474,13 +474,29 @@ def load_scenario(self, explanation_text: str, scenario_id: int):
...
+
+        Args:
+            explanation (str): explanation text
+
+        Returns:
+            str:  generated html explanation
```
Comment (Agrendalath) [inline tests/integration/test_interaction_assessment.py:486]: Nit: isn't this… obvious after reading the first sentence of the docstring? If you want to indicate that it returns a string, please simply add `-> str` after the function name.

**Review by Agrendalath — COMMENTED** [review]
This one looks mostly good. I left few small comments.

File: drag_and_drop_v2/drag_and_drop_v2.py (hunk 3 above)
Comment (abodacs) [reply drag_and_drop_v2/drag_and_drop_v2.py:628]: I followed
https://github.com/openedx/xblock-drag-and-drop-v2/blob/3fddeb57ce28c98d2ad40db6ecb074a69b9e2a0f/drag_and_drop_v2/drag_and_drop_v2.py#L939-L956

as without checking the attr existence raises errors like

```python
AttributeError: 'WorkbenchRuntime' object has no attribute 'replace_course_urls'
````

**Comment by natabene** [comment]
@Agrendalath I personally don't, but maybe @jmbowman wants.

File: drag_and_drop_v2/drag_and_drop_v2.py [hunk 7]
```diff
@@ -620,8 +620,14 @@ def show_answer(self, data, suffix=''):
 
         answer = self._get_correct_state()
 
-        answer['explanation'] = self.data.get('explanation', '').strip()
-
+        explanation = self.data.get('explanation', '').strip()
+        if hasattr(self.runtime, 'replace_urls'):
+            explanation = self.runtime.replace_urls(explanation)
+        if hasattr(self.runtime, 'replace_course_urls'):
+            explanation = self.runtime.replace_course_urls(explanation)
```
Comment (Agrendalath) [inline drag_and_drop_v2/drag_and_drop_v2.py:627]: This is throwing the following exception for me in Studio on Maple:
```python
Internal Server Error: /preview/xblock/block-v1:edX+DemoX+Demo_Course+type@drag-and-drop-v2+block@87425faf77f748d59f2e926c2ac9b17c/handler/show_answer
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/clickjacking.py", line 50, in wrapped_view
    resp = view_func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/preview.py", line 72, in preview_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/mixins.py", line 84, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1506, in handle
    return super().handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/runtime.py", line 1081, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/mixins.py", line 63, in wrapper
    response = func(self, request_json, suffix)
  File "/edx/shared-src/xblock-drag-and-drop-v2/drag_and_drop_v2/drag_and_drop_v2.py", line 627, in show_answer
    explanation = self.runtime.replace_course_urls(explanation)
TypeError: 'NoneType' object is not callable
```

Comment (Agrendalath) [reply drag_and_drop_v2/drag_and_drop_v2.py]: Fixed in 27cc6cc3fe4460af2e2187d5ad4dd5d67b97058f.

**Review by Agrendalath — APPROVED** [review]
:+1:

* [x] I tested this: checked that explanations render HTML correctly; verified Maple and Master compatibility
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

**Comment by e0d** [comment]
Did this change consider implications for cross-site scripting (XSS)?  If we're adding HTML support we should ensure that we're not creating an XSS issues.  There are some useful docs [here](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/preventing_xss/preventing_xss.html#philosophy-and-general-rules).

**Comment by openedx-webhooks** [comment]
@abodacs 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xblock-drag-and-drop-v2&entry.867055334=https://github.com/openedx/xblock-drag-and-drop-v2/pull/295&entry.1484655318=https://github.com/abodacs&entry.752974735=2022-09-24+23:12&entry.1917517419=2022-10-13+18:30&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by openedx-webhooks** [comment]
@abodacs 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xblock-drag-and-drop-v2&entry.867055334=https://github.com/openedx/xblock-drag-and-drop-v2/pull/295&entry.1484655318=https://github.com/abodacs&entry.752974735=2022-09-24+23:12&entry.1917517419=2022-10-13+18:30&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**